File: trunk/ccdream/ccdream/EventDispatcher.m
//
//  EventDispatcher.m
//  ccdream
//
//  Created by wei zhang on 12-6-6.
//  Copyright 2012年 Clican. All rights reserved.
//

#import "EventDispatcher.h"
#import "Session.h"
#import "State.h"
#import "Event.h"
#import "EngineContext.h"
#import "Variable.h"
#import "IState.h"

@implementation EventDispatcher

static EventDispatcher *sharedEventDispatcher = nil;

+(EventDispatcher*) sharedEventDispatcher{
    @synchronized(self) {
		if (sharedEventDispatcher == nil){
            sharedEventDispatcher = [[self alloc] init]; // assignment not done here
        }
	}
	return sharedEventDispatcher;
}


-(void) dispatch:(long) sessionId forState:(long) stateId forEventType:(NSString*) eventType forParameters:(NSDictionary*) parameters{
    @try {
        EngineContext* engineContext = [EngineContext sharedEngineContext];
        
        Event* event = [[[Event alloc] init] autorelease];
        event.completeTime = [NSDate date];
        State* activeState = nil;
        if(stateId==0){
             activeState = [engineContext findActiveStateBySessionId:sessionId];
        }else{
            activeState = [engineContext findStateById:stateId sessionId:sessionId];
        }
        
        if(activeState==nil||![activeState.status isEqualToString:@"active"]){
            [NSException raise:@"Active state not found" format:@"There is no active state for session[%l] state[%l]",sessionId,stateId];
        }
        event.state = activeState;
        event.eventType = eventType;
        if(parameters!=nil){
            for (NSString* name in parameters.allKeys) {
                id value = [parameters objectForKey:name];
                Variable* var = [[[Variable alloc] init] autorelease];
                var.name = name;
                var.value = value;
                var.event = event;
                var.changeDate = [NSDate date];
                [event.variables addObject:var];
            }
        }
        IState* istate = [engineContext getState:activeState.session.name stateName:activeState.name];
        [istate handle:event];
    }
    @catch (NSException *exception) {
        CCLOGERROR(@"handle event exception%@",exception);
        @throw exception;
    }
    @finally {
        
    }
}
@end

File: 2013-05-03-1.1.0/modules/pluto-module-fsm/src/main/java/com/clican/pluto/fsm/engine/EventDispatcher.java
/**
 * The Clican-Pluto software suit is Copyright 2009, Clican Company and individual contributors, and is licensed under the GNU LGPL.
 *
 * @author wei.zhang
 *
 */
package com.clican.pluto.fsm.engine;

import java.io.Serializable;
import java.util.Map;

import com.clican.pluto.fsm.enumeration.EventType;

/**
 * 该接口是工作流引擎的事件驱动接口。
 * <p>
 * 当有新的事件发生的时候一律调用该接口来把事件分发到对应的状态节点上去处理。
 * 
 * @author wei.zhang
 * 
 */
public interface EventDispatcher {

	/**
	 * Dispatch the event.
	 * 
	 * @param sessionId
	 *            The session id
	 * @param stateId
	 *            状态基处理流水号
	 * @param eventType
	 *            事件类型
	 * @param parameters
	 *            该事件相关的参数
	 */
	public void dispatch(Long sessionId, Long stateId, EventType eventType,
			Map<String, Serializable> parameters);
}

// $Id$